### PR TITLE
CRUD on Collection & Community logo

### PR DESCRIPTION
**Pull request by jonas-atmire**
This is the REST API implementation for the rest contract mentioned in:
* https://github.com/DSpace/Rest7Contract/pull/79 
This has recently been slightly altered by the following as well
* https://github.com/DSpace/Rest7Contract/pull/84

Quick overview:
This extends the codebase with the CRUD functionality for the community and collection logos

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java [hunk 1]
```diff
@@ -253,4 +262,28 @@ protected void delete(Context context, UUID id) throws AuthorizeException {
...
+     * Called by request mappings in CollectionRestController
+     * @param context
+     * @param collection    The collection on which to install the logo
+     * @param uploadfile    The new logo
+     * @return              The created bitstream containing the new logo
+     * @throws IOException
+     * @throws AuthorizeException
+     * @throws SQLException
+     */
+    public Bitstream setLogo(Context context, Collection collection, MultipartFile uploadfile)
```
Comment (cwilper) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java]: RestRepositories generally don't return JPA entity objects. They return RestModels. So in this example, it'd be BitstreamRest that would be returned.

**Review by cwilper — CHANGES_REQUESTED** [review]
Looks good in general, but see my comment about RestRepository returning BitstreamRest instead of Bitstream. After addressing that and ensuring tests pass, this seems mergeable.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java (hunk 1 above)
Comment (jonas-atmire) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java]: Great catch @cwilper, I've updated this accordingly to match the other RestRepository methods

**Review by cwilper — APPROVED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/CollectionRestController.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.multipart.MultipartFile;
+
+/**
+ * This RestController takes care of the creation and deletion of Collection's nested objects
+ * This class will typically receive the UUID of a Collection and it'll perform logic on its nested objects
+ */
+@RestController
+@RequestMapping("/api/" + CollectionRest.CATEGORY + "/" + CollectionRest.PLURAL_NAME)
+public class CollectionRestController {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/CollectionRestController.java]: I would recommend renaming this Controller to be `CollectionLogoUploadController` or `CollectionLogoController`.  It's current name is misleading as this Controller does NOT implement the entire `/collections` endpoint...it ONLY implements the Logo upload feature.

Additionally, the `@RequestMapping` should be more specific to the `/logo` path....likely something like 

`@RequestMapping("/api/" + CollectionRest.CATEGORY + "/" + CollectionRest.PLURAL_NAME + "/" + REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID + "/logo")`

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/CommunityRestController.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.multipart.MultipartFile;
+
+/**
+ * This RestController takes care of the creation and deletion of Communities' nested objects
+ * This class will typically receive the UUID of a Community and it'll perform logic on its nested objects
+ */
+@RestController
+@RequestMapping("/api/" + CommunityRest.CATEGORY + "/" + CommunityRest.PLURAL_NAME)
+public class CommunityRestController {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/CommunityRestController.java]: I would recommend renaming this Controller to be `CommunityLogoUploadController` or `CommunityLogoController`.  It's current name is misleading as this Controller does NOT implement the entire `/communities` endpoint...it ONLY implements the Logo upload feature.

Additionally, the `@RequestMapping` should be more specific to the `/logo` path....likely something like 

`@RequestMapping("/api/" + CommunityRest.CATEGORY + "/" + CommunityRest.PLURAL_NAME + "/" + REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID + "/logo")`

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java [hunk 4]
```diff
@@ -260,4 +270,28 @@ protected void delete(Context context, UUID id) throws AuthorizeException {
...
+     * Called by request mappings in CommunityRestController
+     * @param context
+     * @param community     The community on which to install the logo
+     * @param uploadfile    The new logo
+     * @return              The created bitstream containing the new logo
+     * @throws IOException
+     * @throws AuthorizeException
+     * @throws SQLException
+     */
+    public Bitstream setLogo(Context context, Community community, MultipartFile uploadfile)
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java]: This should likely return `BitstreamRest`, just like the `CollectionRestRepository.setLogo()` method.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/CollectionRestController.java (hunk 2 above)
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/CollectionRestController.java]: NOTE: The ITs related to this Controller should also be renamed

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/CommunityRestController.java (hunk 3 above)
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/CommunityRestController.java]: NOTE: The ITs related to this Controller should also be renamed

**Review by tdonohue — COMMENTED** [review]
@jonas-atmire : This code looks good overall, but I have a few Controller renaming requests.  I'd like to avoid creating generically named controllers, as the name therefore implies that it is implementing the entire endpoint (e.g. `/collections` or `/communities`).  Instead, we've recommended keeping Controllers smaller and more specifically named, so that it is clear they only implement a subpath (e.g. `/collections/{uuid}/logo`).  

These changes should be pretty simple to make though, as it's essentially just class renaming.  Other than that, this PR is looking good to me.   See inline comments below

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me now.  Thanks @jonas-atmire & @peter-atmire !

**Pull request merged**